rust-runtime/aws-smithy-compression/src/gzip.rs

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use crate::{Compression, CompressionOptions};
+use aws_smithy_runtime_api::box_error::BoxError;
+use flate2::write::GzEncoder;
+use std::fmt;
+use std::io::prelude::*;
+
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub(crate) struct Gzip {
+    compression: flate2::Compression,
+}
+
+impl Gzip {
+    fn compress_bytes(&self, bytes: &[u8], writer: impl Write) -> Result<(), Error> {
+        let mut encoder = GzEncoder::new(writer, self.compression);
+        encoder.write_all(bytes)?;
+        encoder.try_finish()?;
+
+        Ok(())
+    }
+}
+
+impl Compression for Gzip {
+    fn compress_bytes(&mut self, bytes: &[u8], writer: &mut dyn Write) -> Result<(), BoxError> {
+        Gzip::compress_bytes(self, bytes, writer).map_err(Into::into)
+    }
+}
+
+mod http_body_0_4_x {
+    impl crate::http::http_body_0_4_x::RequestCompressor for super::Gzip {
+        fn header_value(self: Box<Self>) -> http_0_2::HeaderValue {
+            http_0_2::HeaderValue::from_static("gzip")
+        }
+    }
+}
+
+mod http_body_1_x {
+    impl crate::http::http_body_1_x::RequestCompressor for super::Gzip {
+        fn header_value(self: Box<Self>) -> http_1_0::HeaderValue {
+            http_1_0::HeaderValue::from_static("gzip")
+        }
+    }
+}
+
+impl From<&CompressionOptions> for Gzip {
+    fn from(options: &CompressionOptions) -> Self {
+        Gzip {
+            compression: flate2::Compression::new(options.level),
+        }
+    }
+}
+
+impl From<CompressionOptions> for Gzip {
+    fn from(options: CompressionOptions) -> Self {
+        Gzip {
+            compression: flate2::Compression::new(options.level),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+struct Error {
```

Comment:
Seems odd to define an error here when we have an `error` module as well? Is this common style? Is the difference just public errors vs private? It makes me wonder if "compression" errors shouldn't be defined as well and instead of returning `BoxError` we return something more concrete 🤔 , what was the deciding factor for making `compress_bytes(..)` return `BoxError`?

Reply:
I think this is just an oversight on my part but let me think about it for a day.

Reply:
I decided to remove this b/c it wasn't necessary.

As for `why don't we return something more concrete?` I don't think that it would help users much. As I see it, the main benefit of returning an actual type is that users can then match on it. For this sort of user error, I don't expect them to address it programmatically. Rather, they should enter a supported algorithm and then re-compile.